source/client.c

```diff
@@ -274,6 +271,8 @@ static void s_mqtt_client_shutdown(
                     "id=%p: Caller requested disconnect from on_interrupted callback, aborting reconnect",
                     (void *)connection);
                 MQTT_CLIENT_CALL_CALLBACK(connection, on_disconnect);
+                /* Everything is done from the eventloop on the connection, leave the ownsership fully to user. */
+                aws_mqtt_client_connection_release(connection);
```

Comment:
It's scary that this VERY IMPORTANT function is called from mulitple places in this very long and crazy function.

Maybe have 1 set_state() function, and that's where we acquire a reference when exiting DISCONNECTED state, or release the reference when entering DISCONNECTED state?